src/hearing_aid.cpp

```diff
@@ -0,0 +1,449 @@
+        uint8_t res = gatt_client_write_value_of_characteristic_without_response(conn_handle,
+                                                                                chars.vol.value_handle,
+                                                                                sizeof(volume),
+                                                                                (uint8_t*)&volume);
+        if (res != ERROR_CODE_SUCCESS) {
+            LOG_ERROR("%s: Volume: write failed with error: 0x%02x\n", side_str, (unsigned int)res);
+        }
+    }
+}
+
+void HA::send_audio_packet(AudioBuffer::G722Buff& packet)
+{
+    if (state == State::AudioPacketReady) {
+        audio_packet = side() == Side::Left ? packet.l.data() : packet.r.data();
+        LOG_AUDIO("%s: Setting audio packet. Seq Num: %d\n", side_str, (int)audio_packet[0]);
+        ++audio_index;
+        state = State::AudioPacketSending;
+        l2cap_request_can_send_now_event(cid);
+    }
+}
+
+void HA::on_can_send_audio_packet_now()
+{
+    LOG_AUDIO("%s: Sending audio packet. Seq Num: %d\n", side_str, (int)audio_packet[0]);
+    state = State::AudioPacketSent;
+    uint8_t res = l2cap_send(cid, audio_packet, sdu_size_bytes);
+    if (res != ERROR_CODE_SUCCESS) {
+        LOG_ERROR("%s: Error sending audio packet with error code: 0x%02x\n", side_str, (unsigned int)res);
+        state = State::AudioPacketReady;
+    }
+
+}
+
+void HA::on_audio_packet_sent()
+{
+    LOG_AUDIO("%s: Sent audio packet. State: %d. Seq Num: %d\n", side_str, static_cast<int>(state), (int)audio_packet[0]);
+    state = State::AudioPacketReady;
+}
+
+bool HA::is_creating_l2cap_channel()
+{
+    using enum State;
+    return is_any_of(state,
+                     GATTConnected,
+                     UpdateConnParamsStart,
+                     UpdateConnParamsCompleted,
+                     L2ConnStart,
+                     L2ConnCompleted);
+}
+
+bool HA::is_streaming_audio()
+{
+    using enum State;
+    return is_any_of(state, AudioPacketReady, AudioPacketSending, AudioPacketSent, ASPStartOk);
+}
+
+HA::ROP::ROP()
+{}
+
+void HA::ROP::read(const uint8_t* data)
+{
+    std::bitset<8>  device_cap{data[1]};
+    std::bitset<8>  feat_map{data[10]};
+    std::bitset<16> codecs{get_val<uint16_t>(&data[15])};
+
+    version = data[0];
+    side = device_cap[0] ? Side::Right : Side::Left;
+    mode = device_cap[1] ? Mode::Binaural : Mode::Mono;
+    csis_supported = device_cap[2];
+    id.manufacturer_id = get_val<uint16_t>(&data[2]);
+    memcpy(id.unique_id.data(), &data[4], id.unique_id.size());
+    le_coc_supported = feat_map[0];
+    render_delay = get_val<uint16_t>(&data[11]);
+    codec_16khz = codecs[1];
+    codec_24khz = codecs[2];
+}
+
+void HA::ROP::print_values()
+{
+    LOG_INFO("Read Only Properties\n"
+        "  Version:        %d\n"
+        "  Side:           %s\n"
+        "  Mode:           %s\n"
+        "  CSIS:           %s\n"
+        "  Manufacture ID: %xxu\n"
```

Comment:
Did you mean %02hx here?

Reply:
Oh probably. I hate printf format specifiers :(

Reply:
Sorry, that field is two bytes isn't it. you want `%04hx`

Reply:
Sigh... thank you.